util/patch/patch.go

```diff
@@ -0,0 +1,108 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package patch
+
+import (
+	"context"
+	"reflect"
+
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	kerrors "k8s.io/apimachinery/pkg/util/errors"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type Helper struct {
+	client        client.Client
+	before        map[string]interface{}
+	hasStatus     bool
+	beforeStatus  interface{}
+	resourcePatch client.Patch
+	statusPatch   client.Patch
+}
+
+func NewHelper(resource runtime.Object, crClient client.Client) (*Helper, error) {
+	if resource == nil {
+		return nil, errors.Errorf("expected non-nil resource")
+	}
+
+	before, err := runtime.DefaultUnstructuredConverter.ToUnstructured(resource)
+	if err != nil {
+		return nil, err
+	}
+
+	hasStatus := false
+	beforeStatus, ok, err := unstructured.NestedFieldCopy(before, "status")
+	if err != nil {
+		return nil, err
+	}
+	if ok {
+		hasStatus = true
+		unstructured.RemoveNestedField(before, "status")
+	}
+
+	resourceCopy := resource.DeepCopyObject()
+	return &Helper{
+		client:        crClient,
+		before:        before,
+		beforeStatus:  beforeStatus,
+		hasStatus:     hasStatus,
+		resourcePatch: client.MergeFrom(resourceCopy),
+		statusPatch:   client.MergeFrom(resourceCopy),
+	}, nil
+}
+
+func (h *Helper) Patch(ctx context.Context, resource runtime.Object) error {
+	if resource == nil {
+		return errors.Errorf("expected non-nil resource")
```

Comment:
Should we make this error a variable instead?

Reply:
Why?

Reply:
Something like

```
var ErrNilResource = errors.Errorf("expected non-nil resource")
```

Not a huge deal, really minor comment. Given that this is a constant string might make sense.

Reply:
I'd prefer to avoid this unless we have a need for it. See also the section on sentinel error values in https://dave.cheney.net/2019/06/10/constant-time